lib/kube/proxy/single_cert_handler.go

```diff
@@ -0,0 +1,190 @@
+		return trace.AccessDenied(overwriteDeniedMsg)
+	}
+
+	return nil
+}
+
+// singleCertHandler extracts routing information from base64-encoded URL
+// parameters into the current auth user context and forwards the request back
+// to the main router with the path prefix (and its embedded routing parameters)
+// stripped.
+func (f *Forwarder) singleCertHandler() httprouter.Handle {
+	return httplib.MakeHandlerWithErrorWriter(func(w http.ResponseWriter, req *http.Request, p httprouter.Params) (any, error) {
+		teleportCluster, kubeCluster, err := parseRouteFromPath(p)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+
+		userTypeI, err := authz.UserFromContext(req.Context())
+		if err != nil {
+			f.log.WarnContext(req.Context(), "error getting user from context", "error", err)
+			return nil, trace.AccessDenied(accessDeniedMsg)
+		}
+
+		// Insert the extracted routing information from the path into the
+		// identity. Some implementation notes:
+		// - This still relies on RouteToCluster and KubernetesCluster identity
+		//   fields, even though these fields are not part of the TLS identity
+		//   when using path-based routing.
+		// - If the Teleport+Kube cluster names resolve to the local node, these
+		//   values will be used directly in their proper handlers once the
+		//   request is rewritten.
+		// - If the route resolves to a remote node, the identity is encoded (in
+		//   JSON form) into forwarding headers using
+		//   `auth.IdentityForwardingHeaders`. The destination node's auth
+		//   middleware is configured to extract this identity (due to
+		//   EnableCredentialsForwarding) and implicitly trusts this routing
+		//   data, assuming the request originated from a proxy.
+		// - In either case, the destination node is ultimately responsible for
+		//   authorizing the request, and routing information set in the
+		//   identity should not be implicitly trusted. (This was ideally never
+		//   the case, given access to resources could be revoked via roles
+		//   before certs expired.)
+
+		var userType authz.IdentityGetter
+		switch o := userTypeI.(type) {
+		case authz.LocalUser:
+			if err := ensureRouteNotOverwritten(&o.Identity, teleportCluster, kubeCluster); err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			o.Identity.RouteToCluster = teleportCluster
+			o.Identity.KubernetesCluster = kubeCluster
+			userType = o
+		case authz.RemoteUser:
+			if err := ensureRouteNotOverwritten(&o.Identity, teleportCluster, kubeCluster); err != nil {
+				return nil, trace.Wrap(err)
+			}
+
+			o.Identity.RouteToCluster = teleportCluster
+			o.Identity.KubernetesCluster = kubeCluster
+			userType = o
+		default:
+			f.log.WarnContext(req.Context(), "Denying proxy access to unsupported user type", "user_type", logutils.TypeAttr(userTypeI))
+			return nil, trace.AccessDenied(accessDeniedMsg)
+		}
+
+		ctx := authz.ContextWithUser(req.Context(), userType)
+		req = req.WithContext(ctx)
```

Comment:
```suggestion
		req = req.Clone(ctx)
```

 deep clone the r.URL because it's a pointer and it's not copied during req.WithContext otherwise we are introducing a data race.

Reply:
Ah, good catch, it's now deep copied as you suggested.